### PR TITLE
Fix torch.trace backward for non-square matrices

### LINKED ISSUE #171704: `torch.trace` backward produces incorrect gradients for non-square inputs.
### 🐛 Describe the bug

The backward implementation for `trace` in `aten/src/ATen/native/TriangularOps.cpp` (`trace_backward_symint`) computes linear indices for the main diagonal using a fixed step of `m + 1` (where `m = sizes[1]` and `n = sizes[0]`). For non-square inputs with `n > m`, this produces an extra index that wraps into the next row (e.g., `(3, 0)` for a `4x2` input), causing gradients to be written outside the diagonal. This yields incorrect gradients.

### Affected code
- File: `aten/src/ATen/native/TriangularOps.cpp`
- Function: `trace_backward_symint`

Relevant lines as of this report:
```cpp
auto grad_input = at::zeros_symint(sizes[0] * sizes[1], grad.options());
auto indices = at::arange(0, grad_input.numel(), sizes[1] + 1, grad.options().dtype(at::kLong));
// ...
grad_input.index_fill_(0, indices, grad);
```

### Reproduction
Minimal Python repro:

```python
import torch

x = torch.zeros(4, 2, requires_grad=True)
out = torch.trace(x)  # uses main diagonal, length = min(4, 2) = 2
out.backward()
print(x.grad)
```

#### Expected gradient
On the main diagonal only, up to `min(n, m)`:

```text
tensor([[1., 0.],
        [0., 1.],
        [0., 0.],
        [0., 0.]])
```

#### Actual gradient (current implementation)
An extra 1 appears at `(3, 0)` due to the linear index sequence `[0, 3, 6]` for `n=4, m=2, step=m+1=3`:

```text
tensor([[1., 0.],
        [0., 1.],
        [0., 0.],
        [1., 0.]])
```

#### Cross-framework comparison (JAX)
JAX produces the expected gradient for the same non-square input:

```python
import jax
import jax.numpy as jnp

x = jnp.zeros((4, 2))
g = jax.grad(jnp.trace)(x)
print(g)
```

Expected output:

```text
[[1., 0.],
 [0., 1.],
 [0., 0.],
 [0., 0.]]
```

### Versions

Observed on PyTorch 2.7.1, but I believe it still exists in the main branch.

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93

### PR DESCRIPTION
**Pull request by npinto**
`trace_backward_symint` computes diagonal indices via `arange(0, rows*cols, cols+1)`, which overshoots the diagonal for tall matrices (rows > cols) and writes gradient into off-diagonal positions.

Bound the arange stop to `min(M, N) * (cols + 1)`. Same fix applied to the functorch decomposition in `PyTorchOperatorHacks.cpp`. Added non-square sample inputs to `sample_inputs_trace` and a regression test in `test_autograd.py`.

Fixes https://github.com/pytorch/pytorch/issues/171704

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -15595,6 +15595,16 @@ def unwrap(t):
         )
         self.assertEqual(x.grad, 2 * torch.ones_like(x))
 
+    def test_trace_backward_nonsquare_171704(self):
+        # https://github.com/pytorch/pytorch/issues/171704
+        for shape in [(5, 2), (7, 3), (4, 1), (1, 5), (2, 5)]:
```
Comment (Skylion007) [inline test/test_autograd.py:15600]: Slight nit: using self.Subtest context here to detect failures in the future would be nice.

Comment (npinto) [reply test/test_autograd.py:15600]: Good point! Done in 74e285c

**Review by Skylion007 — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]
